### ISSUE TITLE
FR: Add a clean_cmd for necessary pre-cleaning tasks (e.g. sudo)

### ISSUE THREAD

**Issue by patmaddox** [thought-machine/please#3297]
One of my build scripts runs as root using `doas`, and so the resulting files are owned by root and have the `schg` flag set. Removing the files requires me to `doas chflags -R noschg ${outputdir} && doas chown -R $(whoami) ${outputdir}`.

Proposal: add a `clean_cmd` option to rules, that will be run before the built-in command that simply removes files. please would still be responsible for the final removal.

```python
genrule(
...
    clean_cmd = [f"doas chflags -R noschg $(dir {target}) && doas chmod -R 666 $(dir {target}); done"]
...
)
```

**Comment by patmaddox** [comment]
> Notably plz clean wouldn't know to invoke it

As of now, yes. As I understand it, `plz clean` essentially does an `rm -rf` type of thing. But instead it could:

1. iterate over rules with `clean_cmd` and run them
2. do its usual `rm` thing

So really it's a matter of making `clean` aware that commands may have more specialized requirements than a simple `rm`.

>It seems very unusual to me to run individual build actions as root...

Is there a way for a rule to run with elevated permissions? My understanding was that either you can `doas plz ...` or you can have `cmd = ["doas ..."]`. But if I could have e.g.

```python
genrule(
  ...
  doas = True
)
```

Then that would be cool. Although, it still doesn't address the `chflags` issue (more detail on that follows).

Also, I _want_ the final outputs to be owned by the running user, not root. So it would add a whole other element to the mix, like `output_owned_user = True`. Simplest thing is just to stick `doas` in the `cmd` list wherever I need it. I do this because I don't want to run the entire please build as root/

## example: building freebsd-src

You can build as non-root with `-D NO_ROOT`, but it generally works better building source as root.

So I have this rule:

```python
    genrule(
        name = name,
        timeout = 60 * 60 * 4, # 4 hours max
        srcs = {
            "BUILD_SH": "scripts/build.sh",
            "FREEBSD_SRC": f":repo--{clean_version}",
            "SRCCONF": "src.conf",
        },
        cmd = [
            "mkdir _build",
            f"doas env -i PATH=${PATH} BUILDDIR=$(realpath _build) SRCDIR=$(realpath ${SRCS_FREEBSD_SRC}) SRCCONF=$(realpath ${SRCS_SRCCONF}) KERNCONF={kernconf} sh ${SRCS_BUILD_SH} > _build/build.log 2> _build/error.log",
            f"mkdir {name}",
            f"echo {name}:{clean_version} > {name}/VERSION",
            f"cp _build/rel/*.txz _build/rel/MANIFEST _build/*.log {name}/",
            f"cp -Rp _build/pkgbase {name}/",
            "doas chflags -R noschg _build && doas rm -rf _build"
        ],
        outs = [name],
        visibility = ["//infra/images/..."]
    )
```

The very last step is my cleanup step - but if the build fails for some reason, there are a bunch of root files left over. Admittedly I haven't tried putting it first yet - but I was assuming that `clean` happens before the `cmd` steps (at least, that's what makes sense to me and what it seems like from the output).

This is a particularly good example because it's not just an ownership problem. It's also a `chflags` problem. Those flags have to be removed from the files before even root can remove them.

## example: building a zpool

This rule creates a zpool backed by a file, that can be mounted by a VM. The underlying build script requires root privileges.

```python
genrule(
    name = "zdata",
    srcs = {
        "BUILD_ZDATA_SH": "scripts/build_zdata.sh",
        "DISTFILES": glob(["dist/zdata/*"], exclude=["#*", "*~"]),
    },
    cmd = [
        "mkdir _build",
        "doas env -i PATH=${PATH} BUILDDIR=_build DISTDIR=$(realpath $(dir :zdata)/dist/zdata) sh ${SRCS_BUILD_ZDATA_SH} > build.log 2> error.log",
        "cp _build/zdata.zfs _build/zdata.img .",
        "cp build.log zdata.log",
        "cp error.log zdata_error.log",
        "doas rm -rf _build"
    ],
    outs = ["zdata.img", "zdata.zfs", "zdata.log", "zdata_error.log"]
)
```

## comments

You'll note that there's not really anything in the rule definition that implies root is needed. But the common pattern is:

- (please) set up the build sources
- (root) run a build script
- (non-priv) copy the produced assets to the output dir

The interesting thing is that for the assets I'm building - tar files and zpools - all of the ownership information is embedded in the asset. So even though a `.tar` gets copied as `patmaddox`, when I extract it as `root` then all of the resulting files are owned by `root`, `nobody`, etc - whatever ownership the files originally had when tarred.

So the result of this build is:

- a bunch of root-owned files on disk that need to be cleaned
- user-owned outputs

If the build completes successfully, my final cmd cleans everything up so please can perform a clean. But if the build fails, `please clean` will also fail because it doesn't have permissions. Even a `doas please clean` will fail because of the flags issue.

Perhaps the answer is a custom script that takes a target, finds the corresponding tmp dir, and does the chflags / chown. I'll give that a go - but I was thinking it would be good to encode that requirement in the rule definition itself.

Basically, it's 100% possible to write wrapper scripts around please - but a big part of the reason for using please is to have a consistent interface and avoid wrapper scripts in the first place! Please is _amazing_ at being able to wrangle my build scripts into a logical structure. I don't want to have to break that structure just because some built files have high permissions / flags.

**Comment by patmaddox** [comment]
Okay after having spent some more time with please, I think this is no longer necessary. The main reason is that build should be clean - so if it fails for some reason, `plz build --shell` and iterate on it until it's clean (including privileged cleanup steps).

The only time where I think this would come up is if some potentially failing command, like a network connection, occurs after a privileged command. Again, this would be a symptom of a flaky build that needs to be made more resilient.
